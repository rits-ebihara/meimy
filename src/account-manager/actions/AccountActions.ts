import { Alert } from 'react-native';
import { getGenericPassword, setGenericPassword } from 'react-native-keychain';
import { Action, Dispatch } from 'redux';
import ShortId from 'shortid';

import { dateParser } from '../../eim-service/EIMServiceAdapter';
import { getConfig } from '../Config';
import { createInitAccountListState, IAccountListState } from '../states/IAccountLisState';
import { IAccountState } from '../states/IAccountState';
import { asyncLoadAccountListAfterShow } from './AccountListActions';
import { langProfile } from '../../LangProfile';

export const SET_ACCOUNT_ACTION = ShortId();

export interface ISetAccountAction extends Action {
    account: IAccountState;
}

export const createSetAccountAction = (account: IAccountState): ISetAccountAction => {
    return {
        account,
        type: SET_ACCOUNT_ACTION,
    };
};

const config = getConfig();

export const asyncSaveAccountAction
    = async (account: IAccountState, dispatch: Dispatch) => {
        try {
            const { siteName, siteDomain, authType, id, eimToken, lastConnect, userId, password } = account;
            // 拡張した他のプロパティが入らないように精査する
            const saveData: IAccountState = {
                authType,
                eimToken,
                id,
                lastConnect,
                password,
                siteDomain,
                siteName,
                userId,
            };
            // 現在のリストをロード
            const json = await getGenericPassword({ service: config.accountListServiceName });
            let accountList: IAccountListState = createInitAccountListState();
            if (!!json && typeof json !== 'boolean') {
                accountList = JSON.parse(json.password, dateParser) as IAccountListState;
            }
            // IDで検索しなければ追加する
            const existIndex = accountList.accounts.findIndex((a) => a.id === id);
            if (existIndex === -1) {
                accountList.accounts.push(saveData);
            } else {
                accountList.accounts[existIndex] = saveData;
            }
            await setGenericPassword('dummy', JSON.stringify(accountList), { service: config.accountListServiceName });
        } catch (error) {
            Alert.alert('error', langProfile.replaceLang('LK_MSG_saveError'));
            console.error(error);
        }
        // ステートを更新して、一覧に戻ったときに画面の表示が新しくなっているようにする
        await asyncLoadAccountListAfterShow(dispatch);
    };

export const REMOVE_ACCOUNT_ACTION = ShortId();

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IRemoveAccountAction extends Action {
}

export const asyncRemoveAccountAction = async (
    targetId: string, dispatch: Dispatch, onSuccess: () => void) => {
    const save = async () => {
        try {
            // 現在のリストをロード
            const json = await getGenericPassword({ service: config.accountListServiceName });
            let accountList: IAccountListState = createInitAccountListState();
            if (!!json && typeof json !== 'boolean') {
                accountList = JSON.parse(json.password) as IAccountListState;
            } else {
                return;
            }
            accountList.accounts = accountList.accounts.filter((a) => a.id !== targetId);
            await setGenericPassword('dummy', JSON.stringify(accountList), { service: config.accountListServiceName });
        } catch (error) {
            Alert.alert('error', langProfile.replaceLang('LK_MSG_saveError'));
            console.error(error);
        }
    };
    await save();
    // ステートを更新して、一覧に戻ったときに画面の表示が新しくなっているようにする
    await asyncLoadAccountListAfterShow(dispatch);
    onSuccess();
};
