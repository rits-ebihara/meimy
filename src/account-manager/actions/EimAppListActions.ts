// import { Toast } from 'native-base';
import { NavigationScreenProp } from 'react-navigation';
import { Action, Dispatch } from 'redux';
import ShortId from 'shortid';

import { EIMServiceAdapter } from '../../eim-service/EIMServiceAdapter';
import { IEimApp } from '../states/IEimAppListState';
import INavigateController from './INavigateController';

interface ISourceEimApp {
    system: {};
    properties: {
        id: string;
        name: string;
        description: string;
    };
}

interface IGetAppListResponse {
    apps: ISourceEimApp[];
}

export const SET_APP_LIST = ShortId();
export interface ISetAppListAction extends Action {
    appList: IEimApp[];
}
export const createSetAppListAction = (appList: IEimApp[]): ISetAppListAction => {
    return {
        appList,
        type: SET_APP_LIST,
    };
};

export const LOAD_APP_LIST = ShortId();

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface ILoadAppListAction extends Action {
}
export const createLoadAppListAction =
    async (dispatch: Dispatch, navigateActions: INavigateController,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>, onError: () => void): Promise<void> => {
        const { siteDomain, siteName, tokens, appKeyPrefix } = navigateActions.getLinkState();
        if (!siteDomain || !tokens || !appKeyPrefix) { return; }
        dispatch({
            type: LOAD_APP_LIST,
        });
        // アプリ一覧の取得
        await (async () => {
            const sa = new EIMServiceAdapter(siteDomain);
            try {
                const response = await sa.getAppList(tokens);
                const responseObj = JSON.parse(response.body) as IGetAppListResponse;
                const appList = responseObj.apps
                    .filter((a) => a.properties.id.startsWith(appKeyPrefix))
                    .map((sourceApp): IEimApp => {
                        const { properties } = sourceApp;
                        return {
                            appKey: properties.id,
                            appName: properties.name,
                            description: properties.description,
                            siteDomain,
                            siteName: siteName || siteDomain,
                            tokens,
                        };
                    });
                // リストに１件しかない場合
                if (appList.length === 1) {
                    const appKey = appList[0].appKey;
                    navigateActions.openApp({ appKey }, navigation);
                } else {
                    dispatch(createSetAppListAction(appList));
                }
            } catch (e) {
                console.error(e);
                onError();
            }
        })();
    };
