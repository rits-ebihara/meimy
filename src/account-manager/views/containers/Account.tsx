import Clone from 'clone';
import { Button, Card, Container, Content, Icon, Input, Item, Label, Picker, Text, Toast } from 'native-base';
import React, { Component } from 'react';
import { Alert, BackHandler, Platform, StyleSheet, TextStyle, ViewStyle } from 'react-native';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import ShortId from 'shortid';

import { FloatingButton } from '../../../components/FloatingButton/FloatingButton';
import { EIMServiceAdapter } from '../../../eim-service/EIMServiceAdapter';
import { ICombinedNavProps, IProps } from '../../../redux-helper/redux-helper';
import { asyncRemoveAccountAction, asyncSaveAccountAction } from '../../actions/AccountActions';
import navigateController from '../../actions/NavigateActions';
import { createShowWebPageAction } from '../../actions/WebSignInActions';
import { getConfig } from '../../Config';
import { IAccountManagerState } from '../../IAccountManagerState';
import RoutePageNames from '../../RoutePageNames';
import { IAccountListState } from '../../states/IAccountLisState';
import { AuthType, IAccountState } from '../../states/IAccountState';
import { langProfile } from '../../../LangProfile';

const config = getConfig();

interface IDiffState {
    loginResultMessage: string;
    siteNameError: boolean;
    siteDomainError: boolean;
    mode: 'view' | 'edit';
    shownEditMenu: boolean;
    shownLoginDialog: boolean;
}

export interface IAccountProps extends IAccountState {
    accountListState: IAccountListState;
}
export interface IAccountLocaleState extends IAccountState, IDiffState {
}

type localStateType = 'siteName' | 'siteDomain' | 'authType';

// interface INavParam {
//     toggleMode: () => void;
//     getMode: () => 'view' | 'edit';
// }
interface IErrors {
    siteNameErrorMessage: string;
    siteErrorMessage: string;
    hasError: boolean;
}
// eslint-disable-next-line @typescript-eslint/class-name-casing
export class _Account extends Component<ICombinedNavProps<IAccountProps>, IAccountLocaleState> {
    public static navigationOptions = ({ navigation }: NavigationScreenProps) => {
        const { colorPalets } = config;
        return {
            headerRight: (
                <Button
                    transparent
                    onPress={navigation.getParam('remove')}
                    style={navigation.getParam('removeButtonStyle')}>
                    <Text style={{ color: '#fff', fontSize: 16, marginTop: 10 }}>
                        {langProfile.replaceLang('LK_delete')}
                    </Text>
                </Button>
            ),
            headerStyle: {
                backgroundColor: colorPalets.$colorPrimary3,
            },
            headerTintColor: colorPalets.$invertColor,
            headerTitle: langProfile.replaceLang('LK_siteInfo'),
        };
    }
    private backupState: IAccountLocaleState | null = null;
    public constructor(props: ICombinedNavProps<IAccountProps>) {
        super(props);
        BackHandler.addEventListener('hardwareBackPress', this.backPage);
        const { cloneProp, diffState }: { cloneProp: IAccountState; diffState: IDiffState } = this.createState(props);
        this.state = Object.assign(cloneProp, diffState);
        delete (this.state as any).accountListState;
    }
    public render = () => {
        const { theme, colorPalets } = config;
        const { state } = this;
        const errorMessageStyle: TextStyle = {
            color: theme.brandDanger,
            fontSize: 12,
        };
        const style = StyleSheet.create({
            textboxStyle: {
                color: (state.mode === 'view') ? colorPalets.$frontDisabledColor : colorPalets.$frontColor,
            },
        });
        const viewingMenu =
            <FloatingButton key="edit_button" iconName="create" primaryButtonColor={theme.brandPrimary}
                onPress={this.onPressEdit} />;
        const editingMenu = [
            <FloatingButton key="save_button" iconName="md-checkmark" iconType="Ionicons"
                onPress={this.onPressSave} primaryButtonColor={theme.brandSuccess} />,
            <FloatingButton key="cancel" iconName="close" primaryButtonColor={theme.brandDanger}
                onPress={this.onPressCancelEdit} position="bottom-left" />,
        ];
        const { siteNameErrorMessage, siteErrorMessage, hasError }: IErrors = this.createErrorMessages(state);
        return (
            <Container>
                <Content>
                    <Card style={{ paddingBottom: 24 }}>
                        <Item stackedLabel>
                            <Label>{langProfile.replaceLang('LK_siteName')}</Label>
                            <Input key="site_name" value={state.siteName} editable={state.mode === 'edit'}
                                style={style.textboxStyle}
                                onChangeText={(text) => { this.onChangeState('siteName', text); }} />
                        </Item>
                        <Text style={errorMessageStyle}>{siteNameErrorMessage}</Text>
                        <Item stackedLabel>
                            <Label>{langProfile.replaceLang('LK_siteDomain')}</Label>
                            <Input key="site_domain" value={state.siteDomain} editable={state.mode === 'edit'}
                                style={style.textboxStyle}
                                keyboardType="url"
                                autoCapitalize="none"
                                onChangeText={(text) => { this.onChangeState('siteDomain', text); }} />
                        </Item>
                        <Text style={errorMessageStyle}>{siteErrorMessage}</Text>
                        <Item style={{ marginTop: 12, marginBottom: 12 }}>
                            <Label>{langProfile.replaceLang('LK_authenticationMethod')}</Label>
                            <Picker
                                key="auth-type"
                                mode="dropdown"
                                enabled={state.mode === 'edit'}
                                iosIcon={<Icon name="ios-arrow-down" />}
                                textStyle={style.textboxStyle as TextStyle}
                                selectedValue={state.authType}
                                style={Platform.OS === 'android' ? style.textboxStyle as TextStyle : {}}
                                onValueChange={(text) => { this.onChangeState('authType', text); }}>
                                <Picker.Item label={langProfile.replaceLang('LK_password')} value="password" />
                                <Picker.Item label={langProfile.replaceLang('LK_office')} value="o365" />
                            </Picker>
                        </Item>
                        {this.createSignInButton(hasError)}
                    </Card>
                </Content>
                {(state.mode === 'view') ? viewingMenu : editingMenu}
            </Container>
        );
    }
    public componentDidMount = () => {
        (async () => { this.inputCheck(); })();
        this.props.navigation.setParams({ remove: this.onPressRemove });
        this.props.navigation.setParams({
            removeButtonStyle: !!this.state.id ? {} : { display: 'none' },
        });
    }
    public componentDidUpdate = (_preProp: ICombinedNavProps<IAccountState>, preState: IAccountLocaleState) => {
        if (this.state.id !== preState.id) {
            this.props.navigation.setParams({
                removeButtonStyle: !!this.state.id ? {} : { display: 'none' },
            });
        }
    }

    private createSignInButton = (hasError: boolean) => {
        const { state } = this;
        const buttonStyle: ViewStyle = {
            marginTop: 24,
        };
        if (state.mode === 'view') {
            return <Button key="sign-in-button"
                block rounded success style={buttonStyle} disabled={hasError}
                onPress={this.onPressConnect}>
                <Text>{langProfile.replaceLang('LK_signIn')}</Text>
            </Button>;
        }
        return null;
    }
    private onPressEdit = () => {
        // キャンセルされたときのために値を保存しておく
        this.backupState = Clone(this.state);
        this.setState({ mode: 'edit' });
    }
    private onPressCancelEdit = () => {
        // 新規作成時の場合は、前の画面に戻る
        if (!this.state.id) {
            this.props.navigation.pop();
            return;
        }
        // 保持しておいた値に戻す
        this.setState(this.backupState);
        this.backupState = null;
    }
    private onPressRemove = () => {
        const me = this;
        Alert.alert(
            langProfile.replaceLang('LK_accountDelete'),
            langProfile.replaceLang('LK_MSG_siteNameDelete', this.state.siteName),
            [
                {
                    style: 'default',
                    text: langProfile.replaceLang('LK_cancel'),
                },
                {
                    onPress: () => {
                        asyncRemoveAccountAction(
                            me.state.id as string,
                            me.props.dispatch,
                            me.navPop);
                    },
                    style: 'destructive',
                    text: langProfile.replaceLang('LK_delete'),
                },
            ],
        );
    }
    private navPop = () => {
        this.props.navigation.pop();
    }
    private onPressConnect = async () => {
        // トークンがあればそれを検証
        const client = new EIMServiceAdapter(this.state.siteDomain);
        if (this.state.eimToken.length !== 0) {
            const tokenResult = await client.validateToken(this.state.eimToken);
            if (tokenResult) {
                await this.successConnect(this.state.eimToken);
                return;
            }
        }
        // if (this.state.authType === 'password') {
        //     this.setState({ shownLoginDialog: true });
        // } else {
        this.props.dispatch(createShowWebPageAction(this.state));
        this.props.navigation.navigate(RoutePageNames.webSignInPageName);
        // }
    }

    private changeInputState: { [key in localStateType]: (value: string) => void } = {
        siteName: (value: string) => {
            this.setState({ siteName: value });
        },
        siteDomain: (value: string) => {
            this.setState({ siteDomain: value });
        },
        authType: (value: string) => {
            this.setState({ authType: value as AuthType });
        },
    };
    private onChangeState = (key: localStateType, value: string) => {
        this.changeInputState[key](value);
        setTimeout(this.inputCheck, 100);
    }
    private onPressSave = async () => {
        if (!this.inputCheck()) {
            return;
        }
        const account = Clone(this.state) as IAccountState;
        if (!account.id) {
            account.id = ShortId();
        }
        // サイトが変更されることもあるので、トークン、ユーザー、パスワードは削除する
        account.eimToken = [];
        account.password = undefined;
        account.userId = undefined;
        // データを保存
        await asyncSaveAccountAction(account, this.props.dispatch);
        this.setState(account);
        this.setState({ mode: 'view' });
    }
    private inputCheck = (state?: IAccountLocaleState) => {
        const targetState = state || this.state;
        let result = true;
        const changeState = {
            siteDomainError: false,
            siteNameError: false,
        };
        // サイト名 必須
        if (!targetState.siteName) {
            result = false;
            changeState.siteNameError = true;
        }
        // ドメイン名 正規表現に一致していること
        if (!/^([a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9]\.){1,}[a-zA-Z]{2,}$/.test(targetState.siteDomain)) {
            result = false;
            changeState.siteDomainError = true;
        }
        // 指定されたドメインのみに限ること
        if (!/(ope\.azure\.ricoh-eim\.com)|(eim.ricoh.com)$/.test(targetState.siteDomain)) {
            result = false;
            changeState.siteDomainError = true;
        }
        this.setState(changeState);
        return result;
    }
    private backPage = () => {
        this.props.navigation.pop();
        return true;
    }

    private createErrorMessages(state: Readonly<IAccountLocaleState>): IErrors {
        const hasError = state.siteDomainError || state.siteNameError;
        const siteErrorMessage: string = (this.state.siteDomainError) ?
            langProfile.replaceLang('LK_MSG_siteDomain') : '';
        const siteNameErrorMessage: string = (this.state.siteNameError) ?
            langProfile.replaceLang('LK_MSG_siteNameRequired') : '';
        return { siteNameErrorMessage, siteErrorMessage, hasError };
    }

    private async successConnect(tokens: string[]) {
        Toast.show({
            text: langProfile.replaceLang('LK_MSG_authentication'),
            type: 'success'
        });
        // トークンを保存する
        const account = Clone(this.state) as IAccountState;
        account.eimToken = tokens;
        await asyncSaveAccountAction(account, this.props.dispatch);
        this.setState(account);
        this.setState({ mode: 'view' });

        // 遷移先を決定する
        await navigateController.navigateForLink(
            this.props.state.accountListState,
            {
                siteDomain: this.state.siteDomain,
                siteName: this.state.siteName,
                tokens,
            }, this.props.dispatch, this.props.navigation);
    }

    private createState(props: ICombinedNavProps<IAccountState>) {
        const cloneProp = Clone(props.state);
        const diffState: IDiffState = {
            loginResultMessage: '',
            mode: 'view',
            shownEditMenu: false,
            shownLoginDialog: false,
            siteDomainError: true,
            siteNameError: true,
        };
        if (!cloneProp.id) {
            diffState.mode = 'edit';
        }
        return { cloneProp, diffState };
    }
}

const mapStateToProps = (state: IAccountManagerState): IProps<IAccountProps> => {
    return {
        state: {
            ...state.account,
            accountListState: state.accountList
        },
    };
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const __private__ = {
    mapStateToProps,
}

export default connect(mapStateToProps)(_Account);
