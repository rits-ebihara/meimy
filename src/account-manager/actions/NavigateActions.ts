import Clone from 'clone';
import { Linking } from 'react-native';
import { NavigationScreenProp } from 'react-navigation';
import { Dispatch } from 'redux';
import UrlParse from 'url-parse';

import { EIMServiceAdapter } from '../../eim-service/EIMServiceAdapter';
import { getConfig } from '../Config';
import { getEimAccount } from '../EimAccount';
import { IAuthAppQuery } from '../IEimAccount';
import RoutePageNames from '../RoutePageNames';
import { IAccountListState } from '../states/IAccountLisState';
import { createInitAccountState } from '../states/IAccountState';
import { IAuthState } from '../states/IAuthStates';
import { createSetAccountAction } from './AccountActions';
import { createSetAuthState } from './AccountListActions';
import { createSetAppListAction } from './EimAppListActions';
import INavigateController from './INavigateController';

export class NavigateController implements INavigateController {
    public parentMainPage?: string;
    public parentNavParams?: { [key: string]: any };
    private linkStates: IAuthState = {};
    private transferAccountPage = (
        accountListState: IAccountListState,
        dispatch: Dispatch,
        navigation: NavigationScreenProp<any>,
        replace: boolean,
        siteDomain: string,
    ) => {
        const accountList = accountListState.accounts;
        let account = accountList.find((a) => a.siteDomain === this.linkStates.siteDomain);
        if (!account) {
            // なければアカウント新規作成画面に遷移する
            account = createInitAccountState();
            const name = siteDomain;
            account.siteName = name;
            account.siteDomain = name;
        }
        dispatch(createSetAccountAction(account));
        if (replace) {
            navigation.replace(RoutePageNames.accountPageName);
        } else {
            navigation.navigate(RoutePageNames.accountPageName);
        }
    };
    private determinedAppAndDomain = async (
        siteDomain: string,
        navigation: NavigationScreenProp<any>,
        accountListState: IAccountListState,
        dispatch: Dispatch,
        replace: boolean,
    ): Promise<boolean> => {
        // トークンがある
        const sa = new EIMServiceAdapter(siteDomain);
        const connected = await sa.validateToken(this.linkStates.tokens || []);
        // トークン検証
        if (!!connected) {
            // 成功
            // 呼び出しアプリを起動
            await this.openApp({}, navigation);
            return true;
        } else {
            // 失敗
            // アカウント画面に遷移する
            this.transferAccountPage(
                accountListState, dispatch, navigation, replace, siteDomain);
            return true;
        }
    }
    private deteminedAppOnly = async (
        domain: string,
        replace: boolean,
        navigation: NavigationScreenProp<any>,
        dispatch: Dispatch,
        accountListState: IAccountListState,
    ) => {
        const sa = new EIMServiceAdapter(domain);
        // 接続済みの場合
        // トークンがある場合は、検証する
        const connected = await sa.validateToken(this.linkStates.tokens || []);
        // 検証の結果
        if (connected) {
            // 認証が成功の場合
            // アプリ一覧を取得
            // アプリ一覧の画面に遷移する
            const moveTo = (replace) ? navigation.replace : navigation.navigate;
            if (this.linkStates.appKeyPrefix) {
                dispatch(createSetAppListAction([]));
                moveTo(RoutePageNames.appListPageName);
            } else {
                moveTo(RoutePageNames.accountListPageName);
            }
            return true;
        } else {
            // 失敗の場合、アカウントの画面に遷移する
            this.transferAccountPage(
                accountListState, dispatch, navigation, replace, domain);
        }
        return true;
    }
    public navigateForLink = async (
        accountListState: IAccountListState,
        pLinkState: IAuthState,
        dispatch: Dispatch,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>,
        replace: boolean = false
    ): Promise<boolean> => {
        this.linkStates = Object.assign({}, this.linkStates, pLinkState);
        if (!!this.linkStates.siteDomain && !!this.linkStates.appKey) {
            // ドメインとアプリが決まっている
            return await this.determinedAppAndDomain(
                this.linkStates.siteDomain,
                navigation,
                accountListState,
                dispatch,
                replace);
        } else if (!!this.linkStates.siteDomain) {
            // サイトドメインが決まっている
            // トークンがある
            return await this.deteminedAppOnly(
                this.linkStates.siteDomain,
                replace,
                navigation,
                dispatch,
                accountListState,
            );
        } else {
            // ドメインの指定がない場合は、なにもしない
            return false;
        }
    }
    public clear = () => {
        this.linkStates = {};
    }
    public getLinkState = () => {
        return Clone(this.linkStates);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public openApp = async (pLinkState: IAuthState, navigation: NavigationScreenProp<any>) => {
        this.linkStates = Object.assign({}, this.linkStates, pLinkState);
        const eimAccount = getEimAccount();
        eimAccount.appKey = this.linkStates.appKey || '';
        eimAccount.domain = this.linkStates.siteDomain || '';
        eimAccount.siteName = this.linkStates.siteName || '';
        eimAccount.eimTokens = this.linkStates.tokens || [];
        await eimAccount.save();
        await eimAccount.loadUser();
        if (this.parentMainPage) {
            navigation.navigate(this.parentMainPage, this.parentNavParams);
        }
    }
    public openAccountManager = async (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>,
        dispatch: Dispatch,
        link?: string,
        hash?: string
    ) => {
        let appKey: string | undefined;
        let domain: string | undefined;
        // 部分適用
        const open = (appKey?: string, domain?: string) => {
            return this._openAccountManager(
                dispatch, navigation, link, hash,
                appKey, domain);
        }
        // メールなど文書を開くリンクから開く
        if (!!link && !!hash) {
            const linkUrl = UrlParse(link);
            domain = linkUrl.host;
            const paths = hash.split('/');
            appKey = paths.length === 5 ? paths[2] : undefined;
            open(appKey, domain);
            return;
        }
        const eimAccount = getEimAccount();
        // 前回アクセスしたサイトを取得する
        await eimAccount.load().then((lastAccount) => {
            // 認証アプリを起動する
            if (!!lastAccount) {
                open(lastAccount.appKey, lastAccount.domain);
            } else {
                open();
            }
        });
    }
    private _openAccountManager = async (
        dispatch: Dispatch, navigation: NavigationScreenProp<any>,
        link?: string, hash?: string,
        appKey?: string, domain?: string
    ) => {
        const config = getConfig();

        const authState: IAuthState = {
            appKeyPrefix: config.appKeyPrefix,
        };
        const accountManagerUrl = UrlParse('eimapplink-accountmanager://accountmanager/');
        const query: IAuthAppQuery = {
            appprefix: config.appKeyPrefix,
            mapp: config.appKeyPrefix,
        };
        if (!!appKey && !!domain) {
            query.appkey = appKey;
            query.domain = domain;
            authState.appKey = appKey;
            authState.siteDomain = domain;
        }
        if (!!link) {
            const linkUrl = UrlParse(link);
            linkUrl.set('hash', hash);
            query.link = linkUrl.href;
            authState.link = linkUrl.href;
        }

        accountManagerUrl.set('query', query);
        await Linking.canOpenURL(accountManagerUrl.href).then((result) => {
            if (result) {
                Linking.openURL(accountManagerUrl.href);
            } else {
                this.parentMainPage = config.startPage;
                if (!!hash) {
                    const paths = hash.split('/');
                    this.parentNavParams = {
                        parameter: paths.length === 5 ? paths[4] : undefined,
                    };
                }
                dispatch(createSetAuthState(authState));
                navigation.navigate(RoutePageNames.authPageName);
            }
        });
    }
}
export default new NavigateController();
