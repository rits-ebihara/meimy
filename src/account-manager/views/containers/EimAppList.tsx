import { Body, Container, ListItem, Spinner, Text, Toast } from 'native-base';
import React, { Component } from 'react';
import { BackHandler } from 'react-native';
import { connect } from 'react-redux';

import { ICombinedNavProps, IProps } from '../../../redux-helper/redux-helper';
import { createLoadAppListAction, createSetAppListAction } from '../../actions/EimAppListActions';
import NavigateActions from '../../actions/NavigateActions';
import { getConfig } from '../../Config';
import { IAccountManagerState } from '../../IAccountManagerState';
import IEimAppList from '../../states/IEimAppListState';

const config = getConfig();

class EimAppList extends Component<ICombinedNavProps<IEimAppList>> {
    public static navigationOptions = () => {
        return {
            headerStyle: {
                backgroundColor: config.colorPalets.$colorPrimary3,
            },
            headerTintColor: config.colorPalets.$invertColor,
            headerTitle: 'アプリ一覧',
        };
    }
    public constructor(props: ICombinedNavProps<IEimAppList>) {
        super(props);
        BackHandler.addEventListener('hardwareBackPress', this.onPressBackButton);
    }
    public render = () => {
        const appListBox = this.props.state.appList.map((app) => {
            return (
                <ListItem noIndent key={app.appKey} onPress={this.onPressItem.bind(this, app.appKey)}>
                    <Body>
                        <Text>{app.appName}</Text>
                        <Text note>{app.description}</Text>
                    </Body>
                </ListItem>
            );
        });
        return (
            <Container>
                {this.props.state.loading ? <Spinner color="blue" /> : appListBox}
            </Container>
        );
    }
    public componentDidMount = () => {
        createLoadAppListAction(this.props.dispatch, NavigateActions, this.props.navigation, this.onNetworkError);
    }
    public onPressItem = (appKey: string) => {
        NavigateActions.openApp({ appKey }, this.props.navigation);
    }

    private onNetworkError = () => {
        Toast.show({
            text: 'ネットワークエラーが発生しました。',
            type: 'warning',
        });
        this.props.dispatch(createSetAppListAction([]));
    }

    private onPressBackButton = () => {
        this.props.navigation.pop();
        return true;
    }
}

const mapStateToProps = (state: IAccountManagerState): IProps<IEimAppList> => {
    return {
        state: state.appList,
    };
};

export const __private__ = {
    EimAppList: EimAppList,
    mapStateToProps,
};

export default connect(mapStateToProps)(EimAppList);
