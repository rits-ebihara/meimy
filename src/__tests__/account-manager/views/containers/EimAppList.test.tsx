import Enzyme, { shallow } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Toast } from 'native-base';
import React from 'react';

import { createLoadAppListAction, createSetAppListAction } from '../../../../account-manager/actions/EimAppListActions';
import NavigateActions from '../../../../account-manager/actions/NavigateActions';
import IEimAppList, { IEimApp } from '../../../../account-manager/states/IEimAppListState';
import { __private__ } from '../../../../account-manager/views/containers/EimAppList';
import { ICombinedNavProps } from '../../../../redux-helper/redux-helper';

Enzyme.configure({ adapter: new EnzymeAdapter() });


jest.mock('native-base', () => ({
    ...jest.requireActual('native-base'),
    Toast: {
        show: jest.fn(),
    }
}));
jest.mock('../../../../account-manager/actions/EimAppListActions', () => {
    return {
        createLoadAppListAction: jest.fn(),
        createSetAppListAction: jest.fn(),
    }
});

jest.mock('../../../../account-manager/actions/NavigateActions', () => {
    return {
        openApp: jest.fn(),
    };
});

const getProps = (): ICombinedNavProps<IEimAppList> => {
    return {
        dispatch: jest.fn(),
        navigation: {
            pop: jest.fn(),
        } as any,
        state: {
            appList: [],
            loading: false,
        },
    };
};

const createAppList = (): IEimApp[] => {
    return [
        {
            appKey: 'appkey001',
            appName: 'app001',
            description: 'app001-desc',
            siteDomain: 'site-domain-001',
            siteName: 'site001',
            tokens: ['abc', 'def'],
        },
        {
            appKey: 'appkey002',
            appName: 'app002',
            description: 'app002-desc',
            siteDomain: 'site-domain-002',
            siteName: 'site002',
            tokens: ['abc', 'def'],
        },
    ];
};

describe('init', () => {
    test('empty list', () => {
        const props = getProps();
        const wrapper = shallow(<__private__.EimAppList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(createLoadAppListAction).toBeCalled();
        expect(createSetAppListAction).not.toBeCalled();
    });
    test('exist list', () => {
        const props = getProps();
        props.state.appList = createAppList();
        const wrapper = shallow(<__private__.EimAppList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    test('loading', () => {
        const props = getProps();
        props.state.appList = createAppList();
        props.state.loading = true;
        const wrapper = shallow(<__private__.EimAppList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('events', () => {
    let wrapper: Enzyme.ShallowWrapper;
    let instance: any;
    let props: ICombinedNavProps<IEimAppList>;
    beforeEach(() => {
        props = getProps();
        wrapper = shallow(<__private__.EimAppList {...props} />);
        instance = wrapper.instance() as any;
    });
    test('press row item', () => {
        instance['onPressItem']('app001');
        expect(NavigateActions.openApp).toBeCalled();
    });
    test('network error handler', () => {
        instance['onNetworkError']();
        expect(Toast.show).toBeCalledWith({
            text: 'ネットワークエラーが発生しました。',
            type: 'warning',
        });
        expect(createSetAppListAction).toBeCalledWith([]);
        expect(props.dispatch).toBeCalled();
    });
    test('on press back button', () => {
        const result = instance['onPressBackButton']();
        expect(props.navigation.pop).toBeCalled();
        expect(result).toBeTruthy();
    });
})

describe('navigationOptions', () => {
    test('', () => {
        const options = __private__.EimAppList.navigationOptions();
        expect(options).toEqual({
            headerStyle: {
                backgroundColor: '#162756',
            },
            headerTintColor: '#fff',
            headerTitle: 'アプリ一覧',
        });
    })
});

describe('mapStateToProps', () => {
    test('', () => {
        const state = {
            appList: createAppList(),
        };
        const result = __private__.mapStateToProps(state as any);
        expect(result).toStrictEqual({
            state: state.appList,
        });
    });
});
