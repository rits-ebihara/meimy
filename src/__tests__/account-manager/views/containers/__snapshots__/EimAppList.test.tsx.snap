// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init empty list 1`] = `<Styled(Container) />`;

exports[`init exist list 1`] = `
<Styled(Container)>
  <Styled(ListItem)
    key="appkey001"
    noIndent={true}
    onPress={[Function]}
  >
    <Styled(Body)>
      <Styled(Text)>
        app001
      </Styled(Text)>
      <Styled(Text)
        note={true}
      >
        app001-desc
      </Styled(Text)>
    </Styled(Body)>
  </Styled(ListItem)>
  <Styled(ListItem)
    key="appkey002"
    noIndent={true}
    onPress={[Function]}
  >
    <Styled(Body)>
      <Styled(Text)>
        app002
      </Styled(Text)>
      <Styled(Text)
        note={true}
      >
        app002-desc
      </Styled(Text)>
    </Styled(Body)>
  </Styled(ListItem)>
</Styled(Container)>
`;

exports[`init loading 1`] = `
<Styled(Container)>
  <Styled(Spinner)
    color="blue"
  />
</Styled(Container)>
`;
